### PR TITLE
Properly support inline asm indirect output operands

### PR DESCRIPTION
**Pull request by Amanieu**
This PR reverts #29543 and instead implements proper support for "=_m" and "+_m" indirect output operands. This provides a framework on top of which support for plain memory operands ("m", "=m" and "+m") can be implemented.

This also fixes the liveness analysis pass not handling read/write operands correctly.

### CODE REVIEW

**Comment by Amanieu** [comment]
rebased, should work now

**Comment by Amanieu** [comment]
rebased

**Comment by frewsxcv** [comment]
@pnkfelix Have any thoughts about this?

File: src/librustc_front/hir.rs [hunk 1]
```diff
@@ -888,7 +888,7 @@ pub enum Ty_ {
 pub struct InlineAsm {
     pub asm: InternedString,
     pub asm_str_style: StrStyle,
-    pub outputs: Vec<(InternedString, P<Expr>, bool)>,
+    pub outputs: Vec<(InternedString, P<Expr>, bool, bool)>,
```
Comment (eddyb) [inline src/librustc_front/hir.rs]: Could we use an enum here instead of two bools?

**Comment by Amanieu** [comment]
Updated to use a struct instead of a tuple for asm output operands

**Comment by pnkfelix** [comment]
Sorry for delay; unfortunately I can't guarantee I'll have time for this until a week from Monday.  (I might get to it, but there's a lot of other things on my plate in short term)

**Comment by pnkfelix** [comment]
(Darn pokey mobile phone interface...)

**Comment by pnkfelix** [comment]
Okay, this seems fine, since asm is still unstable (and thus I don't mind landing extensions like this to it, especially since it is arguably a bug fix).

For anyone else reading this PR, the `*` constraint is specific to LLVM inline asm, AFAICT.  See docs here: http://llvm.org/docs/LangRef.html#indirect-inputs-and-outputs

**Pull request merged**